src/mon/MgrMonitor.cc

```diff
@@ -588,6 +591,11 @@ bool MgrMonitor::prepare_beacon(MonOpRequestRef op)
       pending_map.clients = clients;
       updated = true;
     }
+  } else if (m->get_available()) {
+    dout(4) << "mgr thinks it is active but it is not, dropping!" << dendl;
+    auto m = make_message<MMgrMap>(MgrMap::create_null_mgrmap());
+    mon.send_reply(op, m.detach());
+    goto out;
```

Comment:
Hrm, at least based on code flow, `false` can be returned from `prepare_beacon` in this case now.  Shouldn't that be `true` since the message is handled?  Setting `updated` to `true` here wouldn't match "Track whether we modified pending_map" definition given in the comment, but ultimately this is an implementation of `PaxosService::prepare_update`:
```
   * @returns 'true' if the update message was handled (e.g., a command that
   *	      went through); 'false' otherwise.
```

Reply:
That comment is wrong. The boolean is used to determine if a proposal of pending is required.

The "was this message handled?" indication is only for `preprocess_beacon`: whether or not the message (beacon) should be forwarded to the leader.

Reply:
Then why `MgrMonitor::prepare_update` returns `true` for unknown message types or when `prepare_command` throws an exception?

Reply:
It should not. Perhaps this is some confusion when writing `PaxosService`caused by the semantics of `preprocess_query`.

Reply:
I've fixed the related code. Good catch.

Reply:
I'm still going through OSDMonitor which is unfortunately more complex to correct.

Reply:
> I'm still going through OSDMonitor which is unfortunately more complex to correct.

I wonder if that (together with `MDSMonitor`, etc) is better separated into another PR?

Reply:
Sure.

Reply:
https://github.com/ceph/ceph/pull/50502
https://github.com/ceph/ceph/pull/50501